cmd/cluster/azure/create.go

```diff
@@ -76,54 +80,91 @@ type AzureEncryptionKey struct {
 	KeyVersion   string
 }
 
-func (o *CreateOptions) Validate(ctx context.Context, opts *core.CreateOptions) error {
+// validatedCreateOptions is a private wrapper that enforces a call of Validate() before Complete() can be invoked.
+type validatedCreateOptions struct {
+	*RawCreateOptions
+}
+
+type ValidatedCreateOptions struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*validatedCreateOptions
+}
+
+func (o *RawCreateOptions) Validate(ctx context.Context, opts *core.CreateOptions) (core.PlatformCompleter, error) {
 	// Check if the network security group is set and the resource group is not
 	if o.NetworkSecurityGroupID != "" && o.ResourceGroupName == "" {
-		return fmt.Errorf("flag --resource-group-name is required when using --network-security-group-id")
+		return nil, fmt.Errorf("flag --resource-group-name is required when using --network-security-group-id")
 	}
 
 	// Validate a resource group is provided when using the disk encryption set id flag
 	if o.DiskEncryptionSetID != "" && o.ResourceGroupName == "" {
-		return fmt.Errorf("--resource-group-name is required when using --disk-encryption-set-id")
+		return nil, fmt.Errorf("--resource-group-name is required when using --disk-encryption-set-id")
 	}
-	return nil
+	return &ValidatedCreateOptions{
+		validatedCreateOptions: &validatedCreateOptions{
+			RawCreateOptions: o,
+		},
+	}, nil
 }
 
-func (o *CreateOptions) Complete(ctx context.Context, opts *core.CreateOptions) error {
-	o.name, o.namespace = opts.Name, opts.Namespace
-	o.externalDNSDomain = opts.ExternalDNSDomain
+// completedCreateOptions is a private wrapper that enforces a call of Complete() before cluster creation can be invoked.
+type completedCreateOptions struct {
+	*ValidatedCreateOptions
+
+	externalDNSDomain string
+	name, namespace   string
+
+	infra         *azureinfra.CreateInfraOutput
+	encryptionKey *AzureEncryptionKey
+	creds         util.AzureCreds
+}
+
+type CreateOptions struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*completedCreateOptions
+}
+
+func (o *ValidatedCreateOptions) Complete(ctx context.Context, opts *core.CreateOptions) (core.Platform, error) {
```

Comment:
Follow up nit:
Any reason why Complete wouldn't be called Default()? not sure why we deviated from validate/default naming.

Also Can we make sure we // document all Public funcs?

Reply:
I haven't seen `Default()`, that's new to me - and in most cases it's *not* defaulting. For instance the cluster creation logic that reaches out and builds infra in the cloud, or reads the infrastructure JSON, that's not defaulting. The majority of what happens in `Complete()` is that kind of thing from what I've seen.